### PR TITLE
server: throw exception if fail to cleanup IP resources when release a public IP

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

When reproduced the issue #8967, I got the following error 

```
2024-05-03T14:24:16,885 WARN  [c.c.n.IpAddressManagerImpl] (API-Job-Executor-44:[ctx-6da31d57, job-15369, ctx-eb716451]) (logid:d43b402b) Unable to revoke all the firewall rules for ip id=2 as a part of ip release

2024-05-03T14:24:29,282 DEBUG [c.c.n.IpAddressManagerImpl] (API-Job-Executor-44:[ctx-6da31d57, job-15369, ctx-eb716451]) (logid:d43b402b) Releasing ip id=2; sourceNat = false

2024-05-03T14:24:29,271 WARN  [c.c.n.IpAddressManagerImpl] (API-Job-Executor-44:[ctx-6da31d57, job-15369, ctx-eb716451]) (logid:d43b402b) Failed to release resources for ip address id=2

2024-05-03T14:24:36,266 WARN  [c.c.n.NetworkServiceImpl] (API-Job-Executor-44:[ctx-6da31d57, job-15369, ctx-eb716451]) (logid:d43b402b) Failed to release public ip address id=2
```

The errors are ignored, public IP is released successfully in cloudstack, but the IP is still associated to a VR. When associate the IP to another network, it caused an issue similar as #8967.
However, the reporter of #8967 could not find any error like "Failed to release" or "Unable to revoke" in their logs, so the root cause of #8967 could be different.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/IpAddressManagerImpl.java [hunk 1]
```diff
@@ -732,8 +732,9 @@ public boolean disassociatePublicIpAddress(long addrId, long userId, Account cal
         PublicIpQuarantine publicIpQuarantine = null;
         // Cleanup all ip address resources - PF/LB/Static nat rules
         if (!cleanupIpResources(addrId, userId, caller)) {
-            success = false;
-            s_logger.warn("Failed to release resources for ip address id=" + addrId);
+            String msg = String.format("Failed to release resources for ip address id=%s", addrId);
+            s_logger.error(msg);
+            throw new CloudRuntimeException(msg);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:737]: does this block any other subsequent calls?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:737]: it does, but I think the process should be stopped here if the error happens.

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

File: server/src/main/java/com/cloud/network/IpAddressManagerImpl.java (hunk 1 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:737]: I think this is very much unreadible code:
- in a method named `disassociatePublicIpAddress(..)`
- after a call to `cleanupIpResources(..)`
- folows a call to `markIpAsUnavailable(..)` 
at the very least the last one must be renamed. I ropose `markIpForReleasing(..)`. Looking in the code, that is what it actually does.

I agree that some action other than marking for releasing should follow a failure to clean up. Let's investigate the callers and see what can be done.

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:737]: Hi @weizhouapache should these be addressed here, or later in separate PR?

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/IpAddressManagerImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:737]: @sureshanaparti @DaanHoogland 
renamed the last method from `markAsUnavailable` to `markAsReleasing`

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/network/IpAddressManagerImpl.java [hunk 2]
```diff
@@ -732,8 +732,9 @@ public boolean disassociatePublicIpAddress(long addrId, long userId, Account cal
         PublicIpQuarantine publicIpQuarantine = null;
         // Cleanup all ip address resources - PF/LB/Static nat rules
         if (!cleanupIpResources(addrId, userId, caller)) {
```
Comment (hsato03) [inline server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:734]: Are the IP resources that are cleaned up by the `cleanupIpResources` method being rolled back when the exception is thrown?

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/network/IpAddressManagerImpl.java:734]: @weizhouapache can you check this ^^^

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by DaanHoogland — COMMENTED** [review]
clgtm, but needs testing

**Comment by kiranchavala** [comment]
@weizhouapache 

I am not able to reproduce the behaviour, do you have any  other specific steps

I have followed these steps 

Create  2 sessions and execute the following API at the same time 

1. Execute disassociate IP address
2. Execute  creates a firewall rule on the same IP

Did not observe the log

egrep 'Failed to release | Unable to revoke ' /var/log/cloudstack/management/management-server.log


The create firewall API returns
`Error: (HTTP 431, error code 4350) Unable to create firewall rule for the IP address ID=5 as IP is not associated with any network and no networkId is passed in`

Also, I see the  IPAddress present in the router even after the disassociate IP address is successfull


cat /etc/cloudstack/ips.json

```
}root@r-4-VM:~# cat /etc/cloudstack/ips.json
{
  "eth0": [
    {
      "add": true,
      "broadcast": "10.1.1.255",
      "cidr": "10.1.1.1/24",
      "device": "eth0",
      "gateway": "",
      "netmask": "255.255.255.0",
      "network": "10.1.1.0/24",
      "nic_dev_id": "0",
      "nw_type": "guest",
      "one_to_one_nat": false,
      "public_ip": "10.1.1.1",
      "size": "24",
      "source_nat": false
    }
  ],
  "eth1": [
    {
      "add": true,
      "broadcast": "169.254.255.255",
      "cidr": "169.254.216.24/16",
      "device": "eth1",
      "gateway": "",
      "netmask": "255.255.0.0",
      "network": "169.254.0.0/16",
      "nic_dev_id": "1",
      "nw_type": "control",
      "one_to_one_nat": false,
      "public_ip": "169.254.216.24",
      "size": "16",
      "source_nat": false
    }
  ],
  "eth2": [
    {
      "add": true,
      "broadcast": "10.0.63.255",
      "cidr": "10.0.54.123/20",
      "device": "eth2",
      "first_i_p": true,
      "gateway": "10.0.48.1",
      "is_private_gateway": false,
      "mtu": "1500",
      "netmask": "255.255.240.0",
      "network": "10.0.48.0/20",
      "new_nic": false,
      "nic_dev_id": 2,
      "nw_type": "public",
      "one_to_one_nat": false,
      "public_ip": "10.0.54.123",
      "size": "20",
      "source_nat": true,
      "vif_mac_address": "1e:00:89:00:00:03"
    },
    {
      "add": false,
      "broadcast": "10.0.63.255",
      "cidr": "10.0.54.124/20",
      "device": "eth2",
      "first_i_p": false,
      "gateway": "10.0.48.1",
      "is_private_gateway": false,
      "mtu": "1500",
      "netmask": "255.255.240.0",
      "network": "10.0.48.0/20",
      "new_nic": false,
      "nic_dev_id": 2,
      "nw_type": "public",
      "one_to_one_nat": false,
      "public_ip": "10.0.54.124",
      "size": "20",
      "source_nat": false,
      "vif_mac_address": "1e:00:89:00:00:03"
    }
  ],
  "id": "ips"
```

**Comment by weizhouapache** [comment]
> @weizhouapache 
> 
> I am not able to reproduce the behaviour, do you have any  other specific steps
> 
> I have followed these steps 
> 
> Create  2 sessions and execute the following API at the same time 
> 
> 1. Execute disassociate IP address
> 2. Execute  creates a firewall rule on the same IP
> 
> Did not observe the log
> 
> egrep 'Failed to release | Unable to revoke ' /var/log/cloudstack/management/management-server.log
> 
> 
> The create firewall API returns
> `Error: (HTTP 431, error code 4350) Unable to create firewall rule for the IP address ID=5 as IP is not associated with any network and no networkId is passed in`
> 
> Also, I see the  IPAddress present in the router even after the disassociate IP address is successfull
> 
> 
> cat /etc/cloudstack/ips.json
> 
> ```
> }root@r-4-VM:~# cat /etc/cloudstack/ips.json
> {
>   "eth0": [
>     {
>       "add": true,
>       "broadcast": "10.1.1.255",
>       "cidr": "10.1.1.1/24",
>       "device": "eth0",
>       "gateway": "",
>       "netmask": "255.255.255.0",
>       "network": "10.1.1.0/24",
>       "nic_dev_id": "0",
>       "nw_type": "guest",
>       "one_to_one_nat": false,
>       "public_ip": "10.1.1.1",
>       "size": "24",
>       "source_nat": false
>     }
>   ],
>   "eth1": [
>     {
>       "add": true,
>       "broadcast": "169.254.255.255",
>       "cidr": "169.254.216.24/16",
>       "device": "eth1",
>       "gateway": "",
>       "netmask": "255.255.0.0",
>       "network": "169.254.0.0/16",
>       "nic_dev_id": "1",
>       "nw_type": "control",
>       "one_to_one_nat": false,
>       "public_ip": "169.254.216.24",
>       "size": "16",
>       "source_nat": false
>     }
>   ],
>   "eth2": [
>     {
>       "add": true,
>       "broadcast": "10.0.63.255",
>       "cidr": "10.0.54.123/20",
>       "device": "eth2",
>       "first_i_p": true,
>       "gateway": "10.0.48.1",
>       "is_private_gateway": false,
>       "mtu": "1500",
>       "netmask": "255.255.240.0",
>       "network": "10.0.48.0/20",
>       "new_nic": false,
>       "nic_dev_id": 2,
>       "nw_type": "public",
>       "one_to_one_nat": false,
>       "public_ip": "10.0.54.123",
>       "size": "20",
>       "source_nat": true,
>       "vif_mac_address": "1e:00:89:00:00:03"
>     },
>     {
>       "add": false,
>       "broadcast": "10.0.63.255",
>       "cidr": "10.0.54.124/20",
>       "device": "eth2",
>       "first_i_p": false,
>       "gateway": "10.0.48.1",
>       "is_private_gateway": false,
>       "mtu": "1500",
>       "netmask": "255.255.240.0",
>       "network": "10.0.48.0/20",
>       "new_nic": false,
>       "nic_dev_id": 2,
>       "nw_type": "public",
>       "one_to_one_nat": false,
>       "public_ip": "10.0.54.124",
>       "size": "20",
>       "source_nat": false,
>       "vif_mac_address": "1e:00:89:00:00:03"
>     }
>   ],
>   "id": "ips"
> ```

@kiranchavala 
Thanks
This might has been fixed by #9234

**Comment by weizhouapache** [comment]
Closing
Reopen if needed

**Pull request closed without merging**